Adding layout tests for postMessage CryptoKey for ServiceWorkers.

nmahendru opened the pull request:
#### fdb77ba046f0ffbbf952879f1f6c7c0c5c4d63e9
<pre>
Adding layout tests for postMessage CryptoKey for ServiceWorkers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=183167">https://bugs.webkit.org/show_bug.cgi?id=183167</a>
<a href="https://rdar.apple.com/124393477">rdar://124393477</a>

Reviewed by Sihui Liu and Alex Christensen.

webCryptoMasterkey in WkWebsiteDataStoreDelegate and webCryptoMasterKeyforWebView in
WKNavigationDelegatePrivate both need to return the same masterKey. This is essential for
both postMessage from MainPage to ServiceWorker and vice versa to work for CryptoKeys.
Here we demonstrate it working via a test.

We can remove the webCryptoMasterKeyForWebView and only have the new one once
Safari can implement webCryptoMasterKey.(Just like in the
test implementation here).

Also, WebKitTestRunner is using the V3Navigation client which is actually using a hardcoded
value for the key for deterministic tests. So Setting the same value for WKWebsiteDataStoreDelegate.

Essentially, ServiceWorkers were enable to do PostMessage of a key but it&apos;s just that after
276088@main it just so happened that the ServiceWorker was using the default key but the main Page
had a hardcoded key. This patch also fixes that.

* LayoutTests/http/tests/workers/service/indexeddb-cryptokey-put-get.https-expected.txt:
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
(async runTest):
(async createKeys): Deleted.
(async testFetchedKeys): Deleted.
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get.js:
(async test):
(async checkKey):
(async createKeys):
(async testKey):
(async event):
(event.catch): Deleted.
* LayoutTests/http/tests/workers/service/resources/sw-test-pre.js:
(bytesToHexString):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate webCryptoMasterKey]):

Canonical link: <a href="https://commits.webkit.org/276197@main">https://commits.webkit.org/276197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f1711639bc84a988f00f55de3583966454c792e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43965 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46402 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46615 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40038 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46269 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26937 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20426 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36264 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44542 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20059 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17282 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17563 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38956 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2025 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48222 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18971 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15545 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20356 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/41847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20565 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6022 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19983 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
```diff
@@ -155,8 +155,10 @@ - (NSData *)webCryptoMasterKey
     // Not so random key
     constexpr size_t keyLength = 16;
-    uint8_t keyBytes[keyLength] = { 1 };
+    // same value as https://github.com/WebKit/WebKit/blob/9854e2932ee221bed4a8260dbc9ec7b0d87189ed/Tools/WebKitTestRunner/TestController.cpp#L133
```
This comment is quite likely to get out of date.

// same value as copyWebCryptoMasterKey

szewai reviewed (commented):
So the problem is client may specify different crypto keys in different delegate methods, which can cause web page and service workers to use different keys.

Instead of relying on clients to provide the same master key in different delegate methods, we should just use one delegate method as the truth: i.e. if both methods are specified, web page should use _WKWebsiteDataStoreDelegate one to be consistent with service worker. 

`WebPageProxy::wrapCryptoKey` could check `m_websiteDataStore->client().webCryptoMasterKey()` first; if it's non-null then use that. Note you will also need to make `WebsiteDataStoreClient::webCryptoMasterKey()` to not return default master key (move the returning default key to the call sites).

We should deprecate the method in NavigationClient as followup, since clients are actually not going to (or supposed to) provide different keys for different pages.

szewai reviewed (approved):
r=me

nmahendru committed:
Adding layout tests for postMessage CryptoKey for ServiceWorkers.
https://bugs.webkit.org/show_bug.cgi?id=183167
rdar://124393477

Reviewed by Sihui Liu and Alex Christensen.

webCryptoMasterkey in WkWebsiteDataStoreDelegate and webCryptoMasterKeyforWebView in
WKNavigationDelegatePrivate both need to return the same masterKey. This is essential for
both postMessage from MainPage to ServiceWorker and vice versa to work for CryptoKeys.
Here we demonstrate it working via a test.

We can remove the webCryptoMasterKeyForWebView and only have the new one once
Safari can implement webCryptoMasterKey.(Just like in the
test implementation here).

Also, WebKitTestRunner is using the V3Navigation client which is actually using a hardcoded
value for the key for deterministic tests. So Setting the same value for WKWebsiteDataStoreDelegate.

Essentially, ServiceWorkers were enable to do PostMessage of a key but it's just that after
276088@main it just so happened that the ServiceWorker was using the default key but the main Page
had a hardcoded key. This patch also fixes that.

* LayoutTests/http/tests/workers/service/indexeddb-cryptokey-put-get.https-expected.txt:
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get-worker.js:
(async runTest):
(async createKeys): Deleted.
(async testFetchedKeys): Deleted.
* LayoutTests/http/tests/workers/service/resources/indexeddb-cryptokey-put-get.js:
(async test):
(async checkKey):
(async createKeys):
(async testKey):
(async event):
(event.catch): Deleted.
* LayoutTests/http/tests/workers/service/resources/sw-test-pre.js:
(bytesToHexString):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate webCryptoMasterKey]):

Canonical link: https://commits.webkit.org/276197@main
